betterproto/__init__.py

```diff
@@ -1030,3 +1032,38 @@ async def _unary_stream(
             await stream.send_message(request, end=True)
             async for message in stream:
                 yield message
+
+    async def _stream_unary(
+        self,
+        route: str,
+        request_iterator: Iterator["IProtoMessage"],
+        request_type: Type[ST],
+        response_type: Type[T],
+    ) -> T:
+        """Make a stream request and return the response."""
+        async with self.channel.request(
+            route, grpclib.const.Cardinality.STREAM_UNARY, request_type, response_type
+        ) as stream:
+            for message in request_iterator:
+                await stream.send_message(message)
+            await stream.send_request(end=True)
+            response = await stream.recv_message()
+            assert response is not None
+            return response
+
+    async def _stream_stream(
+        self,
+        route: str,
+        request_iterator: Iterator["IProtoMessage"],
+        request_type: Type[ST],
+        response_type: Type[T],
+    ) -> AsyncGenerator[T, None]:
+        """Make a stream request and return the stream response iterator."""
+        async with self.channel.request(
+            route, grpclib.const.Cardinality.STREAM_STREAM, request_type, response_type
+        ) as stream:
+            for message in request_iterator:
+                await stream.send_message(message)
```

Comment:
It can't be assumed that the client will want to send all the messages before receiving any. 

Ideally I think this method should:

1. accept `request_iterator: Union[Iterator["IProtoMessage"], AsyncIterator["IProtoMessage", None]]`
2. Return type AsyncIterator[T], since it doesn't actually accept data being sent in
3. if Iterator["IProtoMessage"] is given then wrap it in an AsyncIterator to proceed
4. then iterate to receive data from the caller and send it to the server and to receive the data from he server and yield it concurrently, thus allowing upload/download to occur in any valid order.

Or alternatively it might more elegant for this method to just return an AsyncGenerator, and then it's up to the caller to manage sending and receiving from that generator. This would imply an update to the _unary_stream method for consistency, though it should be possible to keep is backwards compatible. I've not totally thought this through but I think it's worth exploring.

Sorry if you were hoping this would be simpler :P I'd be happy to help iterate towards a better solution.

Reply:
Good suggestions; I'll incorporate these.